# Model Evaluation Exercises

In [13]:
import pandas as pd
import numpy as np
from env import get_connection
import seaborn as sns
import matplotlib.pyplot as plt
from pydataset import data
import os
from scipy import stats

#turn off the pink warning boxes
import warnings
warnings.filterwarnings('ignore')

from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn import metrics

import acquire
import prepare

from env import get_connection
pd.options.display.max_columns = None

%matplotlib inline

1

Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | pred dog   | pred cat   |
|:------------  |-----------:|-----------:|
| actual dog    |         46 |         7  |
| actual cat    |         13 |         34 |


- In the context of this problem, what is a false positive?

- In the context of this problem, what is a false negative? 

- How would you describe this model?

Assuming that dogs are Positive and cats Negative, I would determine here that:
- Actual dog and pred dog is True Negative
- Actual cat pred dog is the False Negative
- Actual cat pred cat is True Negative
- Actual dog pred cat is False Positive



In [1]:
#Class example, where cats are positive
tp = 34 
tn = 46
fp = 7
fn = 13

In [3]:
accuracy = (tp+tn) / (tp+tn+fp+fn)

accuracy

0.8

In [4]:
precision = tp / (tp+fp)
precision

0.8292682926829268

In [5]:
recall = tp / (tp+fn)
recall

0.723404255319149

In [10]:
print(f'Recall is {recall}')
print('Recall is: ', round(recall, 2))
print('Precision is: ', round(precision, 2))
print('Accuracy is: ', round(accuracy, 2))

Recall is 0.723404255319149
Recall is:  0.72
Precision is:  0.83
Accuracy is:  0.8


2 

You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

- Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

- Use the predictions dataset and pandas to help answer the following questions:

In [14]:
codey = pd.read_csv('codey.csv')

In [15]:
codey.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

In [16]:
codey.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

- An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


    -- In this case we would be looking for Precision
    -- I would use the third model for this since it has the highest accuracy when it comes to finding ducks with defects

In [7]:
print(metrics.classification_report(codey['actual'], codey['model1']))

              precision    recall  f1-score   support

      Defect       0.80      0.50      0.62        16
   No Defect       0.96      0.99      0.97       184

    accuracy                           0.95       200
   macro avg       0.88      0.74      0.79       200
weighted avg       0.95      0.95      0.94       200



In [9]:
print(metrics.classification_report(codey['actual'], codey['model2']))

              precision    recall  f1-score   support

      Defect       0.10      0.56      0.17        16
   No Defect       0.94      0.56      0.70       184

    accuracy                           0.56       200
   macro avg       0.52      0.56      0.44       200
weighted avg       0.87      0.56      0.66       200



In [8]:
print(metrics.classification_report(codey['actual'], codey['model3']))

              precision    recall  f1-score   support

      Defect       0.13      0.81      0.23        16
   No Defect       0.97      0.53      0.69       184

    accuracy                           0.56       200
   macro avg       0.55      0.67      0.46       200
weighted avg       0.90      0.56      0.65       200



- Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?


In [19]:
#positive case here = defect

subset = codey[codey.model1 == 'Defect]']

In [28]:
model_precision = (subset.actual == subset.model1).mean()
print('Model 1')
print(f'Model precision: {model_precision.2%}')

SyntaxError: f-string: invalid syntax (627967652.py, line 3)

In [38]:
model_precision = (subset.actual == subset.model2).mean()
print('Model precision: ',{model_precision})

Model precision:  {nan}


     -- I would evaluate this using the Precision rate, since we are trying to determine the accuracy of our capturing of positives, and avoiding false positives
     -- I would use model 1 for this
     
     

3

You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Several models have already been developed with the data, and you can find their results here.

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

In [41]:
paws = pd.read_csv('gives_you_paws.csv')

paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

- In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?


In [25]:
paws['actual'].value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [43]:
paws['baseline'] = paws.actual.value_counts().idxmax()
paws.head()

actual model1 model2 model3 model4 baseline
0    cat    cat    dog    cat    dog      dog
1    dog    dog    cat    cat    dog      dog
2    dog    cat    cat    cat    dog      dog
3    dog    dog    dog    cat    dog      dog
4    cat    cat    cat    dog    dog      dog

In [47]:
(paws.actual == paws.baseline).mean()

0.6508

Baseline accuracy is 65.08%

In [48]:
(paws.actual == paws.model1).mean()

0.8074

Model 1 accuracy is 58.74%

In [49]:
(paws.actual == paws.model2).mean()

0.6304

Model 2 accuracy is 64.24%

In [51]:
(paws.actual == paws.model3).mean()

0.5096

Model 3 accuracy is 50.96%

In [50]:
(paws.actual == paws.model4).mean()

0.7426

Model 4 accuracy is 74.26%

- Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend?


In [52]:
subset1 = paws[paws.model1 == 'dog']
subset2 = paws[paws.model2 == 'dog']
subset3 = paws[paws.model3 == 'dog']
subset4 = paws[paws.model4 == 'dog']

In [55]:
(subset1.actual == subset1.model1).mean()

0.8900238338440586

In [56]:
(subset2.actual == subset2.model2).mean()

0.8931767337807607

In [57]:
(subset3.actual == subset3.model3).mean()

0.6598883572567783

In [58]:
(subset4.actual == subset4.model4).mean()

0.7312485304490948

    -- I would use model 2 for the team that deals with dog pictures, as that model more accurately determines when a picture is a dog.

- Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend?


In [60]:
model_pred = []
for model in paws.columns[1:]:
    subset = paws[paws[model] == 'cat']
    precision = (subset.actual == subset[model]).mean()
    model_pred.append([model,precision])
    
model_pred

[['model1', 0.6897721764420747],
 ['model2', 0.4841220423412204],
 ['model3', 0.358346709470305],
 ['model4', 0.8072289156626506],
 ['baseline', nan]]

In [61]:
pd.DataFrame(model_pred, columns=['model', 'precision'])

model  precision
0    model1   0.689772
1    model2   0.484122
2    model3   0.358347
3    model4   0.807229
4  baseline        NaN

    -- I would use model 4 for the team that uses cat picutres, as that model more accurately guesses when a picture is a cat.

Follow the links below to read the documentation about each function, then apply those functions to the data from the previous problem.

- sklearn.metrics.accuracy_score
- sklearn.metrics.precision_score
- sklearn.metrics.recall_score
- sklearn.metrics.classification_report

- Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recommend? -- With sklearn methods


    -- I would use the 1st model for this, since this model is tied with model 2 for precision but the recall has a higher probability

In [31]:
print(metrics.classification_report(paws['actual'], paws['model1']))

              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [32]:
print(metrics.classification_report(paws['actual'], paws['model2']))

              precision    recall  f1-score   support

         cat       0.48      0.89      0.63      1746
         dog       0.89      0.49      0.63      3254

    accuracy                           0.63      5000
   macro avg       0.69      0.69      0.63      5000
weighted avg       0.75      0.63      0.63      5000



In [33]:
print(metrics.classification_report(paws['actual'], paws['model3']))

              precision    recall  f1-score   support

         cat       0.36      0.51      0.42      1746
         dog       0.66      0.51      0.57      3254

    accuracy                           0.51      5000
   macro avg       0.51      0.51      0.50      5000
weighted avg       0.55      0.51      0.52      5000



In [34]:
print(metrics.classification_report(paws['actual'], paws['model4']))

              precision    recall  f1-score   support

         cat       0.81      0.35      0.48      1746
         dog       0.73      0.96      0.83      3254

    accuracy                           0.74      5000
   macro avg       0.77      0.65      0.66      5000
weighted avg       0.76      0.74      0.71      5000



- Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recommend? -- With sklearn methods


    -- I would use model 4 for the cat team, since it has the highest precision at guessing which picture is a cat